compiler/src/iree/compiler/InputConversion/TOSA/TosaToLinalgExt.cpp

```diff
@@ -0,0 +1,187 @@
+
+      if (indicesTy.getDimSize(0) == 1) {
+        Value zero = builder.create<arith::ConstantOp>(
+            rewriter.getZeroAttr(indicesTy.getElementType()));
+        batchIdx = builder.create<linalg::FillOp>(zero, empty).getResult(0);
+      } else {
+        SmallVector<AffineMap, 3> indexingMaps(
+            2, builder.getMultiDimIdentityMap(indicesTy.getRank()));
+        batchIdx = builder
+                       .create<linalg::GenericOp>(
+                           indicesTy, indices, empty, indexingMaps,
+                           getNParallelLoopsAttrs(indicesTy.getRank()),
+                           [&](OpBuilder &nestedBuilder, Location nestedLoc,
+                               ValueRange blockArgs) {
+                             ImplicitLocOpBuilder b(op.getLoc(), nestedBuilder);
+                             auto index = b.create<linalg::IndexOp>(0);
+                             auto cast = b.create<arith::IndexCastOp>(
+                                 indicesTy.getElementType(), index);
+                             b.create<linalg::YieldOp>(cast.getResult());
+                           })
+                       .getResult(0);
+      }
+
+      indicesTy = indicesTy.clone(
+          {indicesTy.getDimSize(0), indicesTy.getDimSize(1), 2});
+      indices = builder.create<tosa::ConcatOp>(indicesTy,
+                                               ValueRange{batchIdx, indices},
+                                               rewriter.getI64IntegerAttr(2));
+    }
+
+    auto collapseBatch = [](Value value, ImplicitLocOpBuilder &b) -> Value {
+      auto valueTy = value.getType().cast<ShapedType>();
+      llvm::SmallVector<int64_t> collapseShape(valueTy.getShape().drop_front());
+      llvm::SmallVector<ReassociationExprs, 4> collapseMap(valueTy.getRank() -
+                                                           1);
+      collapseMap.front().push_back(b.getAffineDimExpr(0));
+      for (int i = 0; i < collapseMap.size(); i++) {
+        collapseMap[i].push_back(b.getAffineDimExpr(i + 1));
+      }
+
+      int64_t batch = valueTy.getShape().front();
+      int64_t rows = collapseShape.front();
+      int64_t batchDyn = batch == ShapedType::kDynamic;
+      int64_t rowsDyn = rows == ShapedType::kDynamic;
+      collapseShape[0] =
+          (batchDyn || rowsDyn) ? ShapedType::kDynamic : batch * rows;
+
+      return b.create<tensor::CollapseShapeOp>(valueTy.clone(collapseShape),
+                                               value, collapseMap);
+    };
+
+    indices = collapseBatch(indices, builder);
+    updates = collapseBatch(updates, builder);
+
+    // Create the LinalgExt scatter operation.
+    auto scatter = builder.create<IREE::LinalgExt::ScatterOp>(
+        TypeRange{values.getType()}, ValueRange{updates, indices},
+        ValueRange{values}, builder.getDenseI64ArrayAttr({0, 1}),
+        builder.getBoolAttr(true));
+
+    llvm::SmallVector<Location> argLocs(2, op.getLoc());
```

Comment:
I'd probably just elide these, I don't think the argLocs give you anything meaningful given the same as the op.

Reply:
Done.